File: go/test/api_entities_ci_runner_detail_entity_test.go
package sdktest

import (
	"encoding/json"
	"fmt"
	"os"
	"path/filepath"
	"runtime"
	"strings"
	"testing"
	"time"

	sdk "github.com/voxgig-sdk/gitlab-sdk/go"
	"github.com/voxgig-sdk/gitlab-sdk/go/core"

	vs "github.com/voxgig-sdk/gitlab-sdk/go/utility/struct"
)

func TestApiEntitiesCiRunnerDetailEntity(t *testing.T) {
	t.Run("instance", func(t *testing.T) {
		testsdk := sdk.TestSDK(nil, nil)
		ent := testsdk.ApiEntitiesCiRunnerDetail(nil)
		if ent == nil {
			t.Fatal("expected non-nil ApiEntitiesCiRunnerDetailEntity")
		}
	})

	t.Run("basic", func(t *testing.T) {
		setup := api_entities_ci_runner_detailBasicSetup(nil)
		// Per-op sdk-test-control.json skip — basic test exercises a flow
		// with multiple ops; skipping any op skips the whole flow.
		_mode := "unit"
		if setup.live {
			_mode = "live"
		}
		for _, _op := range []string{"update", "load"} {
			if _shouldSkip, _reason := isControlSkipped("entityOp", "api_entities_ci_runner_detail." + _op, _mode); _shouldSkip {
				if _reason == "" {
					_reason = "skipped via sdk-test-control.json"
				}
				t.Skip(_reason)
				return
			}
		}
		// The basic flow consumes synthetic IDs from the fixture. In live mode
		// without an *_ENTID env override, those IDs hit the live API and 4xx.
		if setup.syntheticOnly {
			t.Skip("live entity test uses synthetic IDs from fixture — set GITLAB_TEST_API_ENTITIES_CI_RUNNER_DETAIL_ENTID JSON to run live")
			return
		}
		client := setup.client

		// Bootstrap entity data from existing test data (no create step in flow).
		apiEntitiesCiRunnerDetailRef01DataRaw := vs.Items(core.ToMapAny(vs.GetPath("existing.api_entities_ci_runner_detail", setup.data)))
		var apiEntitiesCiRunnerDetailRef01Data map[string]any
		if len(apiEntitiesCiRunnerDetailRef01DataRaw) > 0 {
			apiEntitiesCiRunnerDetailRef01Data = core.ToMapAny(apiEntitiesCiRunnerDetailRef01DataRaw[0][1])
		}
		// Discard guards against Go's unused-var check when the flow's steps
		// happen not to consume the bootstrap data (e.g. list-only flows).
		_ = apiEntitiesCiRunnerDetailRef01Data

		// UPDATE
		apiEntitiesCiRunnerDetailRef01Ent := client.ApiEntitiesCiRunnerDetail(nil)
		apiEntitiesCiRunnerDetailRef01DataUp0Up := map[string]any{
			"id": apiEntitiesCiRunnerDetailRef01Data["id"],
		}

		apiEntitiesCiRunnerDetailRef01MarkdefUp0Name := "access_level"
		apiEntitiesCiRunnerDetailRef01MarkdefUp0Value := fmt.Sprintf("Mark01-api_entities_ci_runner_detail_ref01_%d", setup.now)
		apiEntitiesCiRunnerDetailRef01DataUp0Up[apiEntitiesCiRunnerDetailRef01MarkdefUp0Name] = apiEntitiesCiRunnerDetailRef01MarkdefUp0Value

		apiEntitiesCiRunnerDetailRef01ResdataUp0Result, err := apiEntitiesCiRunnerDetailRef01Ent.Update(apiEntitiesCiRunnerDetailRef01DataUp0Up, nil)
		if err != nil {
			t.Fatalf("update failed: %v", err)
		}
		apiEntitiesCiRunnerDetailRef01ResdataUp0 := core.ToMapAny(apiEntitiesCiRunnerDetailRef01ResdataUp0Result)
		if apiEntitiesCiRunnerDetailRef01ResdataUp0 == nil {
			t.Fatal("expected update result to be a map")
		}
		if apiEntitiesCiRunnerDetailRef01ResdataUp0["id"] != apiEntitiesCiRunnerDetailRef01DataUp0Up["id"] {
			t.Fatal("expected update result id to match")
		}
		if apiEntitiesCiRunnerDetailRef01ResdataUp0[apiEntitiesCiRunnerDetailRef01MarkdefUp0Name] != apiEntitiesCiRunnerDetailRef01MarkdefUp0Value {
			t.Fatalf("expected %s to be updated, got %v", apiEntitiesCiRunnerDetailRef01MarkdefUp0Name, apiEntitiesCiRunnerDetailRef01ResdataUp0[apiEntitiesCiRunnerDetailRef01MarkdefUp0Name])
		}

		// LOAD
		apiEntitiesCiRunnerDetailRef01MatchDt0 := map[string]any{
			"id": apiEntitiesCiRunnerDetailRef01Data["id"],
		}
		apiEntitiesCiRunnerDetailRef01DataDt0Loaded, err := apiEntitiesCiRunnerDetailRef01Ent.Load(apiEntitiesCiRunnerDetailRef01MatchDt0, nil)
		if err != nil {
			t.Fatalf("load failed: %v", err)
		}
		apiEntitiesCiRunnerDetailRef01DataDt0LoadResult := core.ToMapAny(apiEntitiesCiRunnerDetailRef01DataDt0Loaded)
		if apiEntitiesCiRunnerDetailRef01DataDt0LoadResult == nil {
			t.Fatal("expected load result to be a map")
		}
		if apiEntitiesCiRunnerDetailRef01DataDt0LoadResult["id"] != apiEntitiesCiRunnerDetailRef01Data["id"] {
			t.Fatal("expected load result id to match")
		}

	})
}

func api_entities_ci_runner_detailBasicSetup(extra map[string]any) *entityTestSetup {
	loadEnvLocal()

	_, filename, _, _ := runtime.Caller(0)
	dir := filepath.Dir(filename)

	entityDataFile := filepath.Join(dir, "..", "..", ".sdk", "test", "entity", "api_entities_ci_runner_detail", "ApiEntitiesCiRunnerDetailTestData.json")

	entityDataSource, err := os.ReadFile(entityDataFile)
	if err != nil {
		panic("failed to read api_entities_ci_runner_detail test data: " + err.Error())
	}

	var entityData map[string]any
	if err := json.Unmarshal(entityDataSource, &entityData); err != nil {
		panic("failed to parse api_entities_ci_runner_detail test data: " + err.Error())
	}

	options := map[string]any{}
	options["entity"] = entityData["existing"]

	client := sdk.TestSDK(options, extra)

	// Generate idmap via transform, matching TS pattern.
	idmap := vs.Transform(
		[]any{"api_entities_ci_runner_detail01", "api_entities_ci_runner_detail02", "api_entities_ci_runner_detail03"},
		map[string]any{
			"`$PACK`": []any{"", map[string]any{
				"`$KEY`": "`$COPY`",
				"`$VAL`": []any{"`$FORMAT`", "upper", "`$COPY`"},
			}},
		},
	)

	// Detect ENTID env override before envOverride consumes it. When live
	// mode is on without a real override, the basic test runs against synthetic
	// IDs from the fixture and 4xx's. Surface this so the test can skip.
	entidEnvRaw := os.Getenv("GITLAB_TEST_API_ENTITIES_CI_RUNNER_DETAIL_ENTID")
	idmapOverridden := entidEnvRaw != "" && strings.HasPrefix(strings.TrimSpace(entidEnvRaw), "{")

	env := envOverride(map[string]any{
		"GITLAB_TEST_API_ENTITIES_CI_RUNNER_DETAIL_ENTID": idmap,
		"GITLAB_TEST_LIVE":      "FALSE",
		"GITLAB_TEST_EXPLAIN":   "FALSE",
		"GITLAB_APIKEY":         "NONE",
	})

	idmapResolved := core.ToMapAny(env["GITLAB_TEST_API_ENTITIES_CI_RUNNER_DETAIL_ENTID"])
	if idmapResolved == nil {
		idmapResolved = core.ToMapAny(idmap)
	}

	if env["GITLAB_TEST_LIVE"] == "TRUE" {
		mergedOpts := vs.Merge([]any{
			map[string]any{
				"apikey": env["GITLAB_APIKEY"],
			},
			extra,
		})
		client = sdk.NewGitlabSDK(core.ToMapAny(mergedOpts))
	}

	live := env["GITLAB_TEST_LIVE"] == "TRUE"
	return &entityTestSetup{
		client:        client,
		data:          entityData,
		idmap:         idmapResolved,
		env:           env,
		explain:       env["GITLAB_TEST_EXPLAIN"] == "TRUE",
		live:          live,
		syntheticOnly: live && !idmapOverridden,
		now:           time.Now().UnixMilli(),
	}
}

File: lua/test/api_entities_ci_runner_detail_entity_test.lua
-- ApiEntitiesCiRunnerDetail entity test

local json = require("dkjson")
local vs = require("utility.struct.struct")
local sdk = require("gitlab_sdk")
local helpers = require("core.helpers")
local runner = require("test.runner")

local _test_dir = debug.getinfo(1, "S").source:match("^@(.+/)")  or "./"

describe("ApiEntitiesCiRunnerDetailEntity", function()
  it("should create instance", function()
    local testsdk = sdk.test(nil, nil)
    local ent = testsdk:ApiEntitiesCiRunnerDetail(nil)
    assert.is_not_nil(ent)
  end)

  it("should run basic flow", function()
    local setup = api_entities_ci_runner_detail_basic_setup(nil)
    -- Per-op sdk-test-control.json skip.
    local _live = setup.live or false
    for _, _op in ipairs({"update", "load"}) do
      local _should_skip, _reason = runner.is_control_skipped("entityOp", "api_entities_ci_runner_detail." .. _op, _live and "live" or "unit")
      if _should_skip then
        pending(_reason or "skipped via sdk-test-control.json")
        return
      end
    end
    -- The basic flow consumes synthetic IDs from the fixture. In live mode
    -- without an *_ENTID env override, those IDs hit the live API and 4xx.
    if setup.synthetic_only then
      pending("live entity test uses synthetic IDs from fixture — set GITLAB_TEST_API_ENTITIES_CI_RUNNER_DETAIL_ENTID JSON to run live")
      return
    end
    local client = setup.client

    -- Bootstrap entity data from existing test data.
    local api_entities_ci_runner_detail_ref01_data_raw = vs.items(helpers.to_map(
      vs.getpath(setup.data, "existing.api_entities_ci_runner_detail")))
    local api_entities_ci_runner_detail_ref01_data = nil
    if #api_entities_ci_runner_detail_ref01_data_raw > 0 then
      api_entities_ci_runner_detail_ref01_data = helpers.to_map(api_entities_ci_runner_detail_ref01_data_raw[1][2])
    end

    -- UPDATE
    local api_entities_ci_runner_detail_ref01_ent = client:ApiEntitiesCiRunnerDetail(nil)
    local api_entities_ci_runner_detail_ref01_data_up0_up = {
      id = api_entities_ci_runner_detail_ref01_data["id"],
    }

    local api_entities_ci_runner_detail_ref01_markdef_up0_name = "access_level"
    local api_entities_ci_runner_detail_ref01_markdef_up0_value = "Mark01-api_entities_ci_runner_detail_ref01_" .. tostring(setup.now)
    api_entities_ci_runner_detail_ref01_data_up0_up[api_entities_ci_runner_detail_ref01_markdef_up0_name] = api_entities_ci_runner_detail_ref01_markdef_up0_value

    local api_entities_ci_runner_detail_ref01_resdata_up0_result, err = api_entities_ci_runner_detail_ref01_ent:update(api_entities_ci_runner_detail_ref01_data_up0_up, nil)
    assert.is_nil(err)
    local api_entities_ci_runner_detail_ref01_resdata_up0 = helpers.to_map(api_entities_ci_runner_detail_ref01_resdata_up0_result)
    assert.is_not_nil(api_entities_ci_runner_detail_ref01_resdata_up0)
    assert.are.equal(api_entities_ci_runner_detail_ref01_resdata_up0["id"], api_entities_ci_runner_detail_ref01_data_up0_up["id"])
    assert.are.equal(api_entities_ci_runner_detail_ref01_resdata_up0[api_entities_ci_runner_detail_ref01_markdef_up0_name], api_entities_ci_runner_detail_ref01_markdef_up0_value)

    -- LOAD
    local api_entities_ci_runner_detail_ref01_match_dt0 = {
      id = api_entities_ci_runner_detail_ref01_data["id"],
    }
    local api_entities_ci_runner_detail_ref01_data_dt0_loaded, err = api_entities_ci_runner_detail_ref01_ent:load(api_entities_ci_runner_detail_ref01_match_dt0, nil)
    assert.is_nil(err)
    local api_entities_ci_runner_detail_ref01_data_dt0_load_result = helpers.to_map(api_entities_ci_runner_detail_ref01_data_dt0_loaded)
    assert.is_not_nil(api_entities_ci_runner_detail_ref01_data_dt0_load_result)
    assert.are.equal(api_entities_ci_runner_detail_ref01_data_dt0_load_result["id"], api_entities_ci_runner_detail_ref01_data["id"])

  end)
end)

function api_entities_ci_runner_detail_basic_setup(extra)
  runner.load_env_local()

  local entity_data_file = _test_dir .. "../../.sdk/test/entity/api_entities_ci_runner_detail/ApiEntitiesCiRunnerDetailTestData.json"
  local f = io.open(entity_data_file, "r")
  if f == nil then
    error("failed to read api_entities_ci_runner_detail test data: " .. entity_data_file)
  end
  local entity_data_source = f:read("*a")
  f:close()

  local entity_data = json.decode(entity_data_source)

  local options = {}
  options["entity"] = entity_data["existing"]

  local client = sdk.test(options, extra)

  -- Generate idmap via transform.
  local idmap = vs.transform(
    { "api_entities_ci_runner_detail01", "api_entities_ci_runner_detail02", "api_entities_ci_runner_detail03" },
    {
      ["`$PACK`"] = { "", {
        ["`$KEY`"] = "`$COPY`",
        ["`$VAL`"] = { "`$FORMAT`", "upper", "`$COPY`" },
      }},
    }
  )

  -- Detect ENTID env override before envOverride consumes it. When live
  -- mode is on without a real override, the basic test runs against synthetic
  -- IDs from the fixture and 4xx's. Surface this so the test can skip.
  local entid_env_raw = os.getenv("GITLAB_TEST_API_ENTITIES_CI_RUNNER_DETAIL_ENTID")
  local idmap_overridden = entid_env_raw ~= nil and entid_env_raw:match("^%s*{") ~= nil

  local env = runner.env_override({
    ["GITLAB_TEST_API_ENTITIES_CI_RUNNER_DETAIL_ENTID"] = idmap,
    ["GITLAB_TEST_LIVE"] = "FALSE",
    ["GITLAB_TEST_EXPLAIN"] = "FALSE",
    ["GITLAB_APIKEY"] = "NONE",
  })

  local idmap_resolved = helpers.to_map(
    env["GITLAB_TEST_API_ENTITIES_CI_RUNNER_DETAIL_ENTID"])
  if idmap_resolved == nil then
    idmap_resolved = helpers.to_map(idmap)
  end

  if env["GITLAB_TEST_LIVE"] == "TRUE" then
    local merged_opts = vs.merge({
      {
        apikey = env["GITLAB_APIKEY"],
      },
      extra or {},
    })
    client = sdk.new(helpers.to_map(merged_opts))
  end

  local live = env["GITLAB_TEST_LIVE"] == "TRUE"
  return {
    client = client,
    data = entity_data,
    idmap = idmap_resolved,
    env = env,
    explain = env["GITLAB_TEST_EXPLAIN"] == "TRUE",
    live = live,
    synthetic_only = live and not idmap_overridden,
    now = os.time() * 1000,
  }
end
